src/insexpand.c

```diff
@@ -4330,17 +4366,26 @@ ins_compl_delete(void)
 /*
  * Insert the new text being completed.
  * "in_compl_func" is TRUE when called from complete_check().
+ * "move_cursor" is used when 'completeopt' includes "preinsert" and when TRUE
+ * cursor needs to move back from the inserted text to the compl_leader.
  */
     void
-ins_compl_insert(int in_compl_func)
+ins_compl_insert(int in_compl_func, int move_cursor)
 {
-    int compl_len = get_compl_len();
+    int	    compl_len = get_compl_len();
+    int	    preinsert = ins_compl_has_preinsert();
+    char_u  *str = compl_shown_match->cp_str.string;
+    int	    leader_len = ins_compl_leader_len();
 
     // Make sure we don't go over the end of the string, this can happen with
     // illegal bytes.
     if (compl_len < (int)compl_shown_match->cp_str.length)
-	ins_compl_insert_bytes(compl_shown_match->cp_str.string + compl_len, -1);
-    if (match_at_original_text(compl_shown_match))
+    {
+	ins_compl_insert_bytes(str + compl_len, -1);
+	if (preinsert && move_cursor)
+	    curwin->w_cursor.col -= ((size_t)STRLEN(str) - leader_len);
```

Comment:
Why not use `compl_shown_match->cp_str.length` instead of `STRLEN(str)` here?

Reply:
should using length..

Reply:
Added change to #16542